plutus-core/untyped-plutus-core/src/UntypedPlutusCore/NewParser.hs

```diff
@@ -0,0 +1,139 @@
+
+import qualified NewUntypedPlutusCore            as UPLC
+import qualified PlutusCore                      as PLC
+import qualified PlutusCore.Parsable             as PLC
+import           PlutusPrelude                   (Pretty, through)
+import           Text.Megaparsec                 hiding (ParseError, State, parse)
+import           UntypedPlutusCore.Check.Uniques (checkProgram)
+import           UntypedPlutusCore.Rename        (Rename (rename))
+
+import           Data.ByteString.Lazy            (ByteString)
+import           Data.ByteString.Lazy.Internal   (unpackChars)
+import qualified Data.Text                       as T
+import           PlutusCore.ParserCommon
+
+-- Parsers for UPLC terms
+
+conTerm
+    :: (PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable, PLC.Parsable (PLC.SomeTypeIn (PLC.Kinded uni)))
+    => Parser (UPLC.Term PLC.Name uni fun SourcePos)
+conTerm = inParens $ UPLC.Constant <$> reservedWord "con" <*> constant
+
+builtinTerm :: (Bounded fun, Enum fun, Pretty fun)
+    => Parser (UPLC.Term PLC.Name uni fun SourcePos)
+builtinTerm = inParens $ UPLC.Builtin <$> reservedWord "builtin" <*> builtinFunction
+
+varTerm :: Parser (UPLC.Term PLC.Name uni fun SourcePos)
+varTerm = UPLC.Var <$> getSourcePos <*> name
+
+lamTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+lamTerm tm = inParens $ UPLC.LamAbs <$> reservedWord "lam" <*> name <*> tm
+
+appTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+appTerm tm = inBrackets $ UPLC.Apply <$> getSourcePos <*> tm <*> tm
+
+delayTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+delayTerm tm = inParens $ UPLC.Delay <$> reservedWord "abs" <*> tm
+
+forceTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+forceTerm tm = inBraces $ UPLC.Force <$> getSourcePos <*> tm
+
+errorTerm
+    :: Parser (UPLC.Term PLC.Name uni fun SourcePos)
+errorTerm = inParens $ UPLC.Error <$> reservedWord "error"
+
+-- | Parser for all UPLC terms.
+term
+    :: ( PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable
+       , Bounded fun, Enum fun, Pretty fun, PLC.Parsable (PLC.SomeTypeIn (PLC.Kinded uni)))
+        => Parser (UPLC.Term PLC.Name uni fun SourcePos)
+term = conTerm <|> builtinTerm <|> varTerm
+    <|> lamTerm self <|> appTerm self
```

Comment:
This might be a bit easier to read if each sub-parser was on a line by itself, but it's no big deal.